[SPARK-1600] Refactor FileInputStream tests to remove Thread.sleep() calls and SystemClock usage

JoshRosen opened the pull request:
This patch refactors Spark Streaming's FileInputStream tests to remove uses of Thread.sleep() and SystemClock, which should hopefully resolve some longstanding flakiness in these tests (see SPARK-1600).

Key changes:
- Modify FileInputDStream to use the scheduler's Clock instead of System.currentTimeMillis(); this allows it to be tested using ManualClock.
- Fix a synchronization issue in ManualClock's `currentTime` method.
- Add a StreamingTestWaiter class which allows callers to block until a certain number of batches have finished.
- Change the FileInputStream tests so that files' modification times are manually set based off of ManualClock; this eliminates many Thread.sleep calls.
- Update these tests to use the withStreamingContext fixture.

JoshRosen committed:
Modify FileInputDStream to use Clock class.

JoshRosen committed:
Synchronize `currentTime` in ManualClock

JoshRosen committed:
Add StreamingTestWaiter class.

JoshRosen committed:
Refactor file input stream tests to not rely on SystemClock.

JoshRosen committed:
Remove Thread.sleep calls in FileInputStream CheckpointSuite test.

Hopefully this will fix SPARK-1600.

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/CheckpointSuite.scala:
```diff
@@ -281,102 +279,130 @@ class CheckpointSuite extends TestSuiteBase {
+          newClock.setTime(clock.currentTime())
+          newClock
+        }
+        val waiter = new StreamingTestWaiter(ssc)
```
I dont see the tests in earlier lines 334-336 here. Those test whether after recovering from checkpoint the files are still remember. Although this gets checked eventually by the output, but these test actually fails fast pinpointing the problem. Would be good put them back.

JoshRosen replied:
If you're referring to 334-336 in the original diff, that got rewritten to the more compact

```
 assert(recordedFiles(ssc) === Seq(1, 2, 3))
```

which appears below.  It looks like the test on line 334 checks right after creating the restarted streaming context and I think I've preserved that here on line 363 in the new file.  Am I overlooking something?

tdas replied:
Yeah, you are right! Mind adding a comment on the line saying that so that I dont get confused again :)

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/CheckpointSuite.scala:
```diff
@@ -281,102 +279,130 @@ class CheckpointSuite extends TestSuiteBase {
+        waiter.waitForTotalBatchesStarted(3, batchDuration * 5)
+        Thread.sleep(1000) // To wait for execution to actually begin
+        logInfo("Output = " + outputStream.output.mkString("[", ", ", "]"))
+        assert(outputStream.output.size > 0, "No files processed after restart")
```
I think this error statement is wrong. Its not after restart, its the first start.

JoshRosen replied:
This actually looks like a typo in my refactoring; thanks for catching it.

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/CheckpointSuite.scala:
```diff
@@ -281,102 +279,130 @@ class CheckpointSuite extends TestSuiteBase {
+        clock.addToTime(batchDuration.milliseconds)
+        waiter.waitForTotalBatchesStarted(3, batchDuration * 5)
+        Thread.sleep(1000) // To wait for execution to actually begin
+        logInfo("Output = " + outputStream.output.mkString("[", ", ", "]"))
```
nit: Please rename this "Output after first start = "

JoshRosen replied:
Good idea; I'll take a pass through the logging lines to try to clean them up (I think I pretty much copied them verbatim from the old code).

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/CheckpointSuite.scala:
```diff
@@ -281,102 +279,130 @@ class CheckpointSuite extends TestSuiteBase {
+        }
+        clock.addToTime(batchDuration.milliseconds)
+        waiter.waitForTotalBatchesStarted(3, batchDuration * 5)
+        Thread.sleep(1000) // To wait for execution to actually begin
```
Since the earlier unit test was running in real time, this Thread.sleep was to ensure that the time had progressed since the `ssc.start()` so that the first file that was generated has modtime > context start time. But that is not necessary any more since you are using manual clock and explicitly setting the mod time of the file. So I believe this can be removed.

tdas replied:
Also I think its better to use `waiter.waitForTotalBatchesCompleted()`

tdas replied:
Since we want to make sure that the stop is made before the 3rd batch has completed, lets put an assert after the stop that only 2 batches have been completed (maybe some more method in the waiter) and that the output has only two values (1 and 3).

JoshRosen replied:
I think that my original thought here was that it somehow mattered whether the underlying Spark job had started, with the idea that the batch start event could happen before the corresponding Spark core events.  But from Streaming's perspective, I guess this doesn't really make sense since batch-level metadata and bookkeeping operations would be tied to the Streaming start events and not the Spark ones.

As you've suggested, I can just replace this with an assert on the number of completed batches.

JoshRosen replied:
I tried removing this sleep and it broke the test; I guess that maybe there is a delay between the time that we get the batch started event and the time that the file is recorded.

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/CheckpointSuite.scala:
```diff
@@ -281,102 +279,130 @@ class CheckpointSuite extends TestSuiteBase {
+      // recorded before failure were saved and successfully recovered
+      logInfo("*********** RESTARTING ************")
+      withStreamingContext(new StreamingContext(checkpointDir)) { ssc =>
+        // Copy over the time from the old clock so that we don't appear to have time-traveled:
```
Why `:` at the end?

JoshRosen replied:
Oh, a bad style habit of mine that I still sometimes fall into.  I'll fix this.

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/TestSuiteBase.scala:
```diff
@@ -104,6 +106,77 @@ class TestOutputStreamWithPartitions[T: ClassTag](parent: DStream[T],
 }
 
 /**
+ * This is an interface that can be used to block until certain events occur, such as
```
This is not an interface.

JoshRosen replied:
I'll fix this up, along with my "two spaces after periods" style in the block comments.

tdas commented:
@JoshRosen I did a pass and I had some comments. Overall it looks pretty good. Addressing those comments should make it good enough. 

Could you tell me how much is the speed up by converting the unit tests from real clock to manual clock?

JoshRosen committed:
Merge remote-tracking branch 'origin/master' into SPARK-1600

JoshRosen committed:
Fix log message and comment typos

JoshRosen committed:
Fix log message and comment typos

JoshRosen commented on streaming/src/test/scala/org/apache/spark/streaming/TestSuiteBase.scala:
```diff
@@ -104,6 +106,77 @@ class TestOutputStreamWithPartitions[T: ClassTag](parent: DStream[T],
+  /**
+   * Block until the number of completed batches reaches the given threshold.
+   */
+  def waitForTotalBatchesCompleted(
```
It occurred to me that this might be misleadingly-named since it waits until _at least_ that many batches have been processed.  To avoid this naming issue, plus a proliferation of similar methods, I might be able to just introduce a helper class that encapsulates this "synchronize on an object and wait for a condition involving it to become true" pattern.

I'm imagining that it could look something vaguely like

```scala
def waitUntil[T](obj: T, condition: T => Boolean, timeout: Long): Unit = {
  obj.synchronized {
     while(!condition(obj)) {
      [...] // do the wait() logic here
     }
   }
}

that encapsulates this wait / notify pattern, so could write something like

```scala
waitUntil(waiter, _.completedBatches > 2, timeout, seconds(10)
```

Or, with an implicit conversion, something like

```
waiter.waitUntil(_.completedBatches > 2, timeout=seconds(10))
```

which is a nice-looking syntax and avoids those issues of having to name inequalities.

Similar to [your suggestion](https://github.com/apache/spark/pull/3868/files#r22447509) on another PR, we could add a `pollUntil` method that works for objects that don't support monitor notification / synchronization.

JoshRosen replied:
Actually, something even simpler (since this is just for test code): I can just copy ScalaTest's `eventually` and modify it to use synchronize / notify instead of polling.

JoshRosen replied:
Actually, even this might be overkill: I can just use vanilla `eventually` since we're still blocking on a condition to occur and not relying on real-clock time; I guess using `wait` instead of `sleep` is just an optimization that might save a small amount of test time, but it's not related to flakiness (the goal of this PR).  Therefore, I'll just remove all of this in favor of `eventually` since then I'll benefit from ScalaTest's nice assertion macros.

JoshRosen committed:
Remove last remaining sleep() call

JoshRosen committed:
Revert "Remove last remaining sleep() call"

This reverts commit dbb824700d9de9099d1fba8c96fc8ac6da63b0ae.

JoshRosen committed:
Replace several TestWaiter methods w/ ScalaTest eventually.

JoshRosen committed:
batchTimeToSelectedFiles should be thread-safe

SparkQA commented:
[Test build #25052 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25052/consoleFull) for   PR 3801 at commit [`b4442c3`](https://github.com/apache/spark/commit/b4442c3538ad462a0a7d39f4b2049ed230e92665).
 * This patch merges cleanly.

JoshRosen commented:
Pushed some commits addressing most of the feedback, but I'm still struggling to remove that last `Thread.sleep(1000)`.  I think that the problem here is that the writing of the checkpoint is asynchronous and without the sleep, we wind up in a state where batch 3 has started processing but has not finished, and the StreamingContext shuts down before a snapshot including batch 3's file info is written.  I plan to dig into this tomorrow to see whether this is actually the case.

SparkQA commented:
[Test build #25052 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25052/consoleFull) for   PR 3801 at commit [`b4442c3`](https://github.com/apache/spark/commit/b4442c3538ad462a0a7d39f4b2049ed230e92665).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25052/
Test FAILed.

tdas commented:
Right, that is probably the case. The checkpointing is asynchronous, and there isnt any callback hook for that, which we can use to wait on. We could wait for the expected checkpoint file to appear in the directory. That will make it deterministic.

JoshRosen commented on streaming/src/test/scala/org/apache/spark/streaming/CheckpointSuite.scala:
```diff
@@ -319,102 +318,141 @@ class CheckpointSuite extends TestSuiteBase {
+        // Verify whether all the elements received are as expected
+        val output = outputBuffer.flatMap(x => x)
+        assert(output.contains(6))  // To ensure that the 3rd input (i.e., 3) was processed
+        output.foreach(o =>         // To ensure all the inputs are correctly added cumulatively
```
Can use `toSet` and set comparison methods here.

JoshRosen committed:
Remove Thread.sleep that was used to make task run slowly

JoshRosen committed:
Wait for checkpoint to complete

JoshRosen committed:
Rename StreamingTestWaiter to BatchCounter

JoshRosen committed:
Use standard timeout in ScalaTest `eventually` blocks.

JoshRosen committed:
ConcurrentHashMap -> SynchronizedMap

JoshRosen committed:
Fix some ManualClock usage problems.

JoshRosen committed:
Use set comparisons for output.

SparkQA commented:
[Test build #25071 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25071/consoleFull) for   PR 3801 at commit [`8340bd0`](https://github.com/apache/spark/commit/8340bd0ab50209316202237a9aee7be619a0b922).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25071 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25071/consoleFull) for   PR 3801 at commit [`8340bd0`](https://github.com/apache/spark/commit/8340bd0ab50209316202237a9aee7be619a0b922).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25071/
Test PASSed.

JoshRosen committed:
Address a potential race when setting file modification times

SparkQA commented:
[Test build #25086 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25086/consoleFull) for   PR 3801 at commit [`e4494f4`](https://github.com/apache/spark/commit/e4494f41ed5931c01b2845418e7705dfd19ba9fb).
 * This patch merges cleanly.

tdas commented:
I am merging this. Thanks @JoshRosen for this humongous effort!
